### ISSUE TITLE
Better SciML Training Loop

### ISSUE THREAD

**Issue by ChrisRackauckas** [JuliaDiffEq/DiffEqFlux.jl#120]
I think we should make sure to create some new training loop functions that are more SciML-oriented.

- It should be capable of using Flux and Optim optimizers.
- The same callback should work in both cases.
- It should be about predicting a timeseries and then calculating stuff based on said time series, reusing as much computation as possible.

Following some ideas from @cems2, one idea is the following. 

```julia
function loss_n_ode() 
  timeseries = n_ode(u0)
  sum(abs2,ode_data .- predict_n_ode()),timeseries
end

data = Iterators.repeated((), 1000)
cb = function (loss,timeseries) # cb has all of the returns from the loss!
  display(loss)
  pl = scatter(t,ode_data[1,:],label="data")
  scatter!(pl,t,timeseries[1,:],label="prediction")
  display(plot(pl))
end

# use Flux
DiffEqFlux.sciml_train!(loss_n_ode, Flux.params(u0,dudt), data, ADAM(0.1), cb = cb)
# use Optim
DiffEqFlux.sciml_train!(loss_n_ode, Flux.params(u0,dudt), data, LBFGS(), cb = cb)
```

i.e., setup the callback to use all of the returns from the loss function, where the first return is the
loss. This would dramatically cut down our computation and make it much easier to switch to LBFGS-type methods.

#### Some Notes

@oxinabox mentioned before an example of how to write a Flux `train!` loop, which we can look at the source: 

https://github.com/FluxML/Flux.jl/blob/master/src/optimise/train.jl#L63-L81

@collinwarner found out how to control the Optim optimization state similar to Flux's callbacks, and this can be used to dispatch and build the training loop:

```julia
function cb(opt_state:: Optim.OptimizationState)
    cur_pred = collect(predict_adjoint(opt_state.metadata["x"]))
    n = size(training_data, 1)
    pl = scatter(1:n,training_data[:,10],label="data", legend =:bottomright,title="Spatial Plot at t=$(saveat[10])")
    scatter!(pl,1:n,cur_pred[:,10],label="prediction")
    pl2 = scatter(saveat,training_data[N÷2,:],label="data", legend =:bottomright, title="Timeseries Plot at Middle X")
    scatter!(pl2,saveat,cur_pred[N÷2,:],label="prediction")
    display(plot(pl, pl2, size=(600, 300)))
    display(opt_state.value)
    false
end
cb(trace::Optim.OptimizationTrace) = cb(last(trace))
result =  optimize(loss_adjoint, loss_adjoint_gradient!, pp, BFGS(), Optim.Options(extended_trace=true,callback = cb))
```

**Comment by oxinabox** [comment]
PR to document custom training loops

https://github.com/FluxML/Flux.jl/pull/994

**Comment by ChrisRackauckas** [comment]
A starter PR in this direction is given at https://github.com/JuliaDiffEq/DiffEqFlux.jl/pull/125 . Still needs lots of work, like testing and such, but it might already work.
